common/src/main/java/org/apache/rocketmq/common/topic/TopicValidator.java

```diff
@@ -0,0 +1,131 @@
+    private static final int TOPIC_MAX_LENGTH = 127;
+
+    private static final Set<String> SYSTEM_TOPIC_SET = new HashSet<String>();
+
+    /**
+     * Topics'set which client can not send msg!
+     */
+    private static final Set<String> CLIENT_BLACKLIST_TOPIC_SET = new HashSet<String>();
+
+    static {
+        SYSTEM_TOPIC_SET.add(AUTO_CREATE_TOPIC_KEY_TOPIC);
+        SYSTEM_TOPIC_SET.add(RMQ_SYS_SCHEDULE_TOPIC);
+        SYSTEM_TOPIC_SET.add(RMQ_SYS_BENCHMARK_TOPIC);
+        SYSTEM_TOPIC_SET.add(RMQ_SYS_TRANS_HALF_TOPIC);
+        SYSTEM_TOPIC_SET.add(RMQ_SYS_TRACE_TOPIC);
+        SYSTEM_TOPIC_SET.add(RMQ_SYS_TRANS_OP_HALF_TOPIC);
+        SYSTEM_TOPIC_SET.add(RMQ_SYS_TRANS_CHECK_MAX_TIME_TOPIC);
+        SYSTEM_TOPIC_SET.add(RMQ_SYS_SELF_TEST_TOPIC);
+        SYSTEM_TOPIC_SET.add(RMQ_SYS_OFFSET_MOVED_EVENT);
+
+        CLIENT_BLACKLIST_TOPIC_SET.add(RMQ_SYS_SCHEDULE_TOPIC);
+    }
+
+    private static boolean regularExpressionMatcher(String origin, Pattern pattern) {
+        if (pattern == null) {
+            return true;
+        }
+        Matcher matcher = pattern.matcher(origin);
+        return matcher.matches();
+    }
+
+    public static boolean validateTopic(String topic, RemotingCommand response) {
+
+        if (UtilAll.isBlank(topic)) {
+            response.setCode(ResponseCode.SYSTEM_ERROR);
+            response.setRemark("The specified topic is blank.");
+            return false;
+        }
+
+        if (!regularExpressionMatcher(topic, PATTERN)) {
+            response.setCode(ResponseCode.SYSTEM_ERROR);
+            response.setRemark("The specified topic contains illegal characters, allowing only " + VALID_PATTERN_STR);
+            return false;
+        }
+
+        if (topic.length() > TOPIC_MAX_LENGTH) {
+            response.setCode(ResponseCode.SYSTEM_ERROR);
+            response.setRemark("The specified topic is longer than topic max length.");
+            return false;
+        }
+
+        return true;
+    }
+
+    public static boolean validateSystemTopic(String topic, RemotingCommand response) {
+        if (isSystemTopic(topic)) {
+            response.setCode(ResponseCode.SYSTEM_ERROR);
+            response.setRemark("The topic[" + topic + "] is conflict with system topic.");
+            return false;
+        }
+        return true;
+    }
+
+    public static boolean isSystemTopic(String topic) {
+        return SYSTEM_TOPIC_SET.contains(topic) || topic.startsWith(SYSTEM_TOPIC_PREFIX);
+    }
+
+    public static boolean isBlacklistTopic(String topic) {
+        return CLIENT_BLACKLIST_TOPIC_SET.contains(topic);
+    }
+
+    public static boolean validateBlacklistTopic(String topic, RemotingCommand response) {
+        if (isBlacklistTopic(topic)) {
+            response.setCode(ResponseCode.NO_PERMISSION);
+            response.setRemark("Sending message to topic[" + topic + "] is forbidden.");
+            return false;
+        }
+        return true;
```

Comment:
It would be better to reconsider the naming of these methods. The difference between `validateSystemTopic` and `isSystemTopic` are quite confusing, same for  `isBlacklistTopic` and `validateBlacklistTopic`. For example, when a topic is system topic, but validateSystemTopic return false.